redbot/core/dev_commands.py

```diff
@@ -33,7 +37,272 @@
+        except SyntaxError as exc:
+            self.set_exception(exc, line_offset=1, skip_frames=3)
+            return
+
+        func = self.env["func"]
+        try:
+            with self:
+                self.result = await func()
+        except Exception as exc:
+            self.set_exception(exc, line_offset=1)
+
+    async def run_repl(self) -> None:
+        self.always_include_result = False
+        self._source = self.source
+        executor = None
+        if self.source.count("\n") == 0:
+            # single statement, potentially 'eval'
+            try:
+                code = self.async_compile_with_eval()
+            except SyntaxError:
+                pass
+            else:
+                executor = eval
+
+        if executor is None:
+            try:
+                code = self.async_compile_with_exec()
+            except SyntaxError as exc:
+                self.set_exception(exc, skip_frames=3)
+                return
+
+        try:
+            with self:
+                if executor is None:
+                    result = types.FunctionType(code, self.env)()
+                else:
+                    result = executor(code, self.env)
+                self.result = await maybe_await(result)
+        except Exception as exc:
+            self.set_exception(exc)
+        else:
+            if self.result is not None:
+                self.env["_"] = self.result
+
+    def async_compile_with_exec(self) -> CodeType:
+        return async_compile(self._source, self.filename, "exec")
+
+    def async_compile_with_eval(self) -> CodeType:
+        return async_compile(self._source, self.filename, "eval")
+
+    def format_exception(
+        self, exc: Exception, *, line_offset: int = 0, skip_frames: int = 1
+    ) -> str:
+        """
+        Format an exception to send to the user.
+
+        This function makes a few alterations to the traceback:
+        - First `skip_frames` frames are skipped so that we don't show the frames
+          that are part of Red's code to the user
+        - `FrameSummary` objects that we get from traceback module are updated
+          with the string for the corresponding line of code as otherwise
+          the generated traceback string wouldn't show user's code.
+        - If `line_offset` is passed, this function subtracts it from line numbers
+          in `FrameSummary` objects so that those numbers properly correspond to
+          the code that was provided by the user. This is needed for cases where
+          we wrap user's code in an async function before exec-ing it.
+        """
+        exc_type = type(exc)
+        tb = exc.__traceback__
+        for x in range(skip_frames):
+            if tb is None:
+                break
+            tb = tb.tb_next
+
+        # To mimic linecache module's behavior,
+        # all lines (including the last one) should end with \n.
+        source_lines = [f"{line}\n" for line in self._source.splitlines()]
```

Comment:
```suggestion
        source_lines = source.splitlines(keepends=True)
```

Reply:
Not equivalent, see this commit and the comment above this line:
https://github.com/Cog-Creators/Red-DiscordBot/pull/5843/commits/3877520958c44963e4ed055125d5980fa4907c46

Reply:
The difference can most easily be seen by checking against this test case:
https://github.com/Jackenmen/Red-DiscordBot/blob/f6f7bd5e9e4032aec2c8449b0cebd08f1712bc51/tests/core/test_dev_commands.py#L234-L243
```
    async def _run_dev_output(
        monkeypatch: pytest.MonkeyPatch,
        source: str,
        result: str,
        *,
        debug: bool = False,
        eval: bool = False,
        repl: bool = False,
    ) -> None:
        source = textwrap.dedent(source)
        result = textwrap.dedent(result)
        monkeypatch.setattr("redbot.core.dev_commands.sanitize_output", lambda ctx, s: s)
    
        if debug:
            output = _get_dev_output(source)
            await output.run_debug()
>           assert str(output) == result
E           AssertionError: assert 'Traceback (m...ion by zero\n' == 'Traceback (m...ion by zero\n'
E             Skipping 89 identical leading characters in diff, use -v to show
E               )
E             -         ~~^~~
E             +          ~~^~~
E             ? +
E               ZeroDivisionError: division by zero

tests/core/test_dev_commands.py:162: AssertionError
```